catalog/registry/catalog-registry-federationadminserviceimpl/src/main/java/ddf/catalog/registry/federationadmin/service/impl/FederationAdminServiceImpl.java

```diff
@@ -0,0 +1,305 @@
+
+    private static final String REGISTRY_CONTENT_TYPE = "registry.service";
+
+    private static final String REGISTRY_NODE_PRIMARY = "PRIMARY";
+
+    private CatalogFramework catalogFramework;
+
+    private InputTransformer registryTransformer;
+
+    private Parser parser;
+
+    private ParserConfigurator configurator;
+
+    private final Security security;
+
+    public FederationAdminServiceImpl() {
+        this(Security.getInstance());
+    }
+
+    FederationAdminServiceImpl(Security security) {
+        this.security = security;
+    }
+
+    @Override
+    public void addLocalEntry(Metacard metacard)
+            throws SourceUnavailableException, IngestException {
+        Subject systemSubject = security.getSystemSubject();
+        final CreateRequest createRequest = new CreateRequestImpl(metacard);
+        createRequest.getProperties()
+                .put(SecurityConstants.SECURITY_SUBJECT, systemSubject);
+
+        catalogFramework.create(createRequest);
+    }
+
+    public List<Metacard> getRegistryMetacards()
+            throws UnsupportedQueryException, SourceUnavailableException, FederationException {
+        List<Metacard> registryMetacards = new ArrayList<>();
+
+        Filter filter =
+                FILTER_FACTORY.and(FILTER_FACTORY.like(FILTER_FACTORY.property(Metacard.CONTENT_TYPE),
+                        REGISTRY_CONTENT_TYPE),
+                        FILTER_FACTORY.like(FILTER_FACTORY.property(Metacard.TAGS),
+                                RegistryConstants.REGISTRY_TAG));
+        SortBy sortBy = FILTER_FACTORY.sort(Metacard.CREATED, SortOrder.ASCENDING);
+
+        Query query = new QueryImpl(filter);
+        ((QueryImpl) query).setSortBy(sortBy);
+
+        Subject systemSubject = security.getSystemSubject();
+        final QueryRequest queryRequest = new QueryRequestImpl(query);
+        queryRequest.getProperties()
+                .put(SecurityConstants.SECURITY_SUBJECT, systemSubject);
+
+        QueryResponse queryResponse = catalogFramework.query(queryRequest);
+
+        if (queryResponse != null) {
+            List<Result> results = queryResponse.getResults();
+
+            for (Result result : results) {
+                Metacard metacard = result.getMetacard();
+                if (metacard != null) {
+                    registryMetacards.add(metacard);
+                }
+            }
+        }
+
+        return registryMetacards;
+    }
+
+    public void init() {
+        try {
+            List<Metacard> registryMetacards = getRegistryMetacards();
+            boolean hasIdentityNode = false;
+
+            for (Metacard metacard : registryMetacards) {
+                Attribute regNodeStatus =
+                        metacard.getAttribute(RegistryObjectMetacardType.REGISTRY_NODE_STATUS);
+
+                if (regNodeStatus != null && regNodeStatus.getValue()
+                        .equals("PRIMARY")) {
+                    hasIdentityNode = true;
```

Comment:
Added a filter for the new registry identity node flag and executing a separate query that only returns an identity metacard. No longer looping through entire result set.